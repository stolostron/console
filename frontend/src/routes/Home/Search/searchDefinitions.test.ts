/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import {
  CreateApplicationTopologyLink,
  CreateDetailsLink,
  CreateExternalLink,
  FormatLabels,
  FormatPolicyReportCategories,
  FormatPolicyReportPolicies,
  GetAge,
  GetUrlSearchParam,
  getSearchDefinitions,
} from './searchDefinitions'
import i18next from 'i18next'
const t = i18next.t.bind(i18next)

test('Correctly returns formatSearchbarSuggestions without T in timestamp', () => {
  Date.now = jest.fn(() => 1607028460000)
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
    created: '2020-11-30T14:34:20Z',
  }
  const result = GetAge(item, 'created')
  expect(result).toMatchSnapshot()
})

test('Correctly returns formatSearchbarSuggestions with T in timestamp', () => {
  Date.now = jest.fn(() => 1607028460000)
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
    created: '2020-11-3014:34:20Z',
  }
  const result = GetAge(item, 'created')
  expect(result).toMatchSnapshot()
})

test('Correctly returns formatSearchbarSuggestions no timestamp', () => {
  Date.now = jest.fn(() => 1607028460000)
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
  }
  const result = GetAge(item, 'created')
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateDetailsLink - Cluster', () => {
  const item = {
    name: 'testClusterName',
    namespace: 'testClusterNamespace',
    kind: 'Cluster',
  }
  const result = CreateDetailsLink(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateDetailsLink - ACM-Application', () => {
  const item = {
    name: 'testApplicationName',
    namespace: 'testApplicationNamespace',
    kind: 'Application',
    apigroup: 'app.k8s.io',
    cluster: 'cluster',
    apiversion: 'v1beta1',
  }
  const result = CreateDetailsLink(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateDetailsLink - NON-Application', () => {
  const item = {
    name: 'testApplicationName',
    namespace: 'testApplicationNamespace',
    kind: 'Application',
    cluster: 'testCluster',
    selfLink: '/self/link',
  }
  const result = CreateDetailsLink(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateDetailsLink - HUB-Policy', () => {
  const item = {
    name: 'testPolicyName',
    namespace: 'testPolicyNamespace',
    kind: 'Policy',
    _hubClusterResource: true,
    apigroup: 'policy.open-cluster-management.io',
  }
  const result = CreateDetailsLink(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateDetailsLink - Managed-Policy', () => {
  const item = {
    name: 'testPolicyName',
    namespace: 'testPolicyNamespace',
    kind: 'Policy',
    cluster: 'testCluster',
    selfLink: '/self/link',
  }
  const result = CreateDetailsLink(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateDetailsLink - PolicyReport', () => {
  const item = {
    name: 'testPolicyReport',
    namespace: 'testPolicyReportNamespace',
    kind: 'PolicyReport',
    cluster: 'testCluster',
    selfLink: '/self/link',
  }
  const result = CreateDetailsLink(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateDetailsLink - Default', () => {
  const item = {
    name: 'testPodName',
    namespace: 'testPodNamespace',
    kind: 'Pod',
    cluster: 'testCluster',
    selfLink: '/self/link',
  }
  const result = CreateDetailsLink(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateApplicationTopologyLink', () => {
  const item = {
    name: 'testApp',
    namespace: 'testNamespace',
    kind: 'Application',
    apiversion: 'v1beta1',
    apigroup: 'app.k8s.io',
  }
  const result = CreateApplicationTopologyLink(item, t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns empty CreateApplicationTopologyLink', () => {
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
    dashboard: '',
  }
  const result = CreateApplicationTopologyLink(item, t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateExternalLink from consoleURL', () => {
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
    consoleURL: 'http://consoleurl',
  }
  const result = CreateExternalLink(item, t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns CreateExternalLink from clusterip', () => {
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
    clusterip: 'http://clusterip',
  }
  const result = CreateExternalLink(item, t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns empty CreateExternalLink', () => {
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
  }
  const result = CreateExternalLink(item, t)
  expect(result).toMatchSnapshot()
})

test('Correctly returns label components', () => {
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
    label: 'testlabel=label1; testlabel=label2',
  }
  const result = FormatLabels(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns policyreport policies', () => {
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
    rules:
      'AUTH_OPERATOR_PROXY_ERROR; CONTAINER_ROOT_PARTITION_SIZE; MASTER_DEFINED_AS_MACHINESETS; NODES_MINIMUM_REQUIREMENTS_NOT_MET; UNSUPPORT_SDN_PLUGIN',
    category: 'testcategory=category1; testcategory=category2',
  }
  const result = FormatPolicyReportPolicies(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns category components', () => {
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
    category: 'testcategory=category1; testcategory=category2',
  }
  const result = FormatPolicyReportCategories(item.category)
  expect(result).toMatchSnapshot()
})

test('Correctly returns empty labels', () => {
  const item = {
    name: 'testName',
    namespace: 'testNamespace',
  }
  const result = FormatLabels(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns all resource definitions', () => {
  const testItem = {
    name: 'testName',
    namespace: 'testNamespace',
    kind: 'Pod',
    cluster: 'testCluster',
    label: 'testLabel=label; testLabel1=label1',
    selfLink: '/apigroup/cluster/name',
    created: '2021-01-01T00:00:00Z',
  }
  const searchDefinitions = getSearchDefinitions((key) => key)
  const defKeys = Object.keys(searchDefinitions) as (keyof typeof searchDefinitions)[]
  defKeys.forEach((key) => {
    const definition = searchDefinitions[key].columns.map((col: any) => {
      if (typeof col.cell === 'function') {
        col.cell = col.cell(testItem)
      }
      return col
    })
    expect(definition).toMatchSnapshot(`SearchDefinitions-${key}`)
  })
})

test('Correctly returns url search params with all params & apigroup', () => {
  const item = {
    cluster: 'testCluster',
    kind: 'Pod',
    apiversion: 'v1',
    name: 'testName',
    namespace: 'testNamespace',
  }
  const result = GetUrlSearchParam(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns url search params with all params without apigroup', () => {
  const item = {
    cluster: 'testCluster',
    kind: 'Pod',
    apiversion: 'v1',
    name: 'testName',
    namespace: 'testNamespace',
  }
  const result = GetUrlSearchParam(item)
  expect(result).toMatchSnapshot()
})

test('Correctly returns url search params with 0 params', () => {
  const item = {
    cluster: 'testCluster',
  }
  const result = GetUrlSearchParam(item)
  expect(result).toMatchSnapshot()
})
