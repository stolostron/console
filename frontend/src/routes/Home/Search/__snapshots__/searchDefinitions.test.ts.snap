// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly returns CreateApplicationTopologyLink 1`] = `
<a
  href="/multicloud/applications/details/testNamespace/testApp/topology?apiVersion=application.app.k8s.io"
>
  View topology
</a>
`;

exports[`Correctly returns CreateDetailsLink - ACM-Application 1`] = `
<a
  href="/multicloud/applications/details/testApplicationNamespace/testApplicationName/topology?apiVersion=application.app.k8s.io&cluster=cluster"
>
  testApplicationName
</a>
`;

exports[`Correctly returns CreateDetailsLink - Cluster 1`] = `
<a
  href="/multicloud/infrastructure/clusters/details/testClusterName/overview"
>
  testClusterName
</a>
`;

exports[`Correctly returns CreateDetailsLink - Default 1`] = `
<Link
  to={
    Object {
      "pathname": "/multicloud/home/search/resources",
      "search": "?cluster%3DtestCluster%26kind%3DPod%26namespace%3DtestPodNamespace%26name%3DtestPodName",
      "state": Object {
        "from": "/multicloud/home/search",
        "fromSearch": "",
      },
    }
  }
>
  testPodName
</Link>
`;

exports[`Correctly returns CreateDetailsLink - HUB-Policy 1`] = `
<a
  href="/multicloud/governance/policies/details/testPolicyNamespace/testPolicyName"
>
  testPolicyName
</a>
`;

exports[`Correctly returns CreateDetailsLink - Managed-Policy 1`] = `
<Link
  to={
    Object {
      "pathname": "/multicloud/home/search/resources",
      "search": "?cluster%3DtestCluster%26kind%3DPolicy%26namespace%3DtestPolicyNamespace%26name%3DtestPolicyName",
      "state": Object {
        "from": "/multicloud/home/search",
        "fromSearch": "",
      },
    }
  }
>
  testPolicyName
</Link>
`;

exports[`Correctly returns CreateDetailsLink - NON-Application 1`] = `
<Link
  to={
    Object {
      "pathname": "/multicloud/home/search/resources",
      "search": "?cluster%3DtestCluster%26kind%3DApplication%26namespace%3DtestApplicationNamespace%26name%3DtestApplicationName",
      "state": Object {
        "from": "/multicloud/home/search",
        "fromSearch": "",
      },
    }
  }
>
  testApplicationName
</Link>
`;

exports[`Correctly returns CreateDetailsLink - PolicyReport 1`] = `
<a
  href="/multicloud/infrastructure/clusters/details/testPolicyReportNamespace/overview?showClusterIssues%3Dtrue"
>
  testPolicyReport
</a>
`;

exports[`Correctly returns CreateExternalLink from clusterip 1`] = `"http://clusterip"`;

exports[`Correctly returns CreateExternalLink from consoleURL 1`] = `
<a
  href="http://consoleurl"
  rel="noopener noreferrer"
  target="_blank"
>
  Launch
</a>
`;

exports[`Correctly returns category components 1`] = `
<AcmLabels
  collapse={Array []}
  labels={
    Array [
      "testcategory=category1",
      "testcategory=category2",
    ]
  }
/>
`;

exports[`Correctly returns empty CreateApplicationTopologyLink 1`] = `"-"`;

exports[`Correctly returns empty CreateExternalLink 1`] = `"-"`;

exports[`Correctly returns empty labels 1`] = `"-"`;

exports[`Correctly returns formatSearchbarSuggestions no timestamp 1`] = `"-"`;

exports[`Correctly returns formatSearchbarSuggestions with T in timestamp 1`] = `"3 days ago"`;

exports[`Correctly returns formatSearchbarSuggestions without T in timestamp 1`] = `"3 days ago"`;

exports[`Correctly returns label components 1`] = `
<AcmLabels
  collapse={Array []}
  labels={
    Array [
      "testlabel=label1",
      "testlabel=label2",
    ]
  }
/>
`;

exports[`Correctly returns policyreport policies 1`] = `
<LabelGroup
  aria-label="Label group category"
  categoryName=""
  closeBtnAriaLabel="Close label group"
  collapsedText="3 more"
  defaultIsOpen={false}
  expandedText="Show less"
  hasEditableTextArea={false}
  isClosable={false}
  isCompact={false}
  isEditable={false}
  isVertical={false}
  numLabels={2}
  onClick={[Function]}
  tooltipPosition="top"
>
  <Label
    render={[Function]}
    style={
      Object {
        "backgroundColor": "#fff",
        "padding": "0 .25rem 0 0",
      }
    }
  >
    AUTH_OPERATOR_PROXY_ERROR
  </Label>
  <Label
    render={[Function]}
    style={
      Object {
        "backgroundColor": "#fff",
        "padding": "0 .25rem 0 0",
      }
    }
  >
    CONTAINER_ROOT_PARTITION_SIZE
  </Label>
  <Label
    render={[Function]}
    style={
      Object {
        "backgroundColor": "#fff",
        "padding": "0 .25rem 0 0",
      }
    }
  >
    MASTER_DEFINED_AS_MACHINESETS
  </Label>
  <Label
    render={[Function]}
    style={
      Object {
        "backgroundColor": "#fff",
        "padding": "0 .25rem 0 0",
      }
    }
  >
    NODES_MINIMUM_REQUIREMENTS_NOT_MET
  </Label>
  <Label
    render={[Function]}
    style={
      Object {
        "backgroundColor": "#fff",
        "padding": "0 .25rem 0 0",
      }
    }
  >
    UNSUPPORT_SDN_PLUGIN
  </Label>
</LabelGroup>
`;

exports[`Correctly returns url search params with 0 params 1`] = `"?cluster%3DtestCluster"`;

exports[`Correctly returns url search params with all params & apigroup 1`] = `"?cluster%3DtestCluster%26kind%3DPod%26apiversion%3Dv1%26namespace%3DtestNamespace%26name%3DtestName"`;

exports[`Correctly returns url search params with all params without apigroup 1`] = `"?cluster%3DtestCluster%26kind%3DPod%26apiversion%3Dv1%26namespace%3DtestNamespace%26name%3DtestName"`;
